x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilder.java

```diff
@@ -174,34 +182,28 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
 
     @Override
     protected Query doToQuery(SearchExecutionContext context) throws IOException {
-        if ((pinnedIds != null && pinnedIds.isEmpty() == false) && (pinnedDocs != null && pinnedDocs.isEmpty() == false)) {
-            throw new IllegalArgumentException("applied rules contain both pinned ids and pinned docs, only one of ids or docs is allowed");
-        }
-
-        if (pinnedIds != null && pinnedIds.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedIds.toArray(new String[0]));
+        if (pinnedQueryBuilder != null) {
             return pinnedQueryBuilder.toQuery(context);
-        } else if (pinnedDocs != null && pinnedDocs.isEmpty() == false) {
-            PinnedQueryBuilder pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, pinnedDocs.toArray(new Item[0]));
-            return pinnedQueryBuilder.toQuery(context);
-        } else {
-            return organicQuery.toQuery(context);
         }
-
+        return organicQuery.toQuery(context);
     }
 
-    @SuppressWarnings("unchecked")
     @Override
-    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
-        if (pinnedIds != null || pinnedDocs != null) {
-            return this;
-        } else if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        if (pinnedIdsSupplier != null || pinnedDocsSupplier != null) {
             List<String> identifiedPinnedIds = pinnedIdsSupplier != null ? pinnedIdsSupplier.get() : null;
             List<Item> identifiedPinnedDocs = pinnedDocsSupplier != null ? pinnedDocsSupplier.get() : null;
-            if (identifiedPinnedIds == null && identifiedPinnedDocs == null) {
-                return this; // not executed yet
+            if ((identifiedPinnedIds != null && identifiedPinnedIds.isEmpty())
+                && (identifiedPinnedDocs != null && identifiedPinnedDocs.isEmpty())) {
+                return organicQuery;
+            } else if (identifiedPinnedIds != null && identifiedPinnedIds.isEmpty() == false) {
+                pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, identifiedPinnedIds.toArray(new String[0]));
+                return pinnedQueryBuilder;
+            } else if (identifiedPinnedDocs != null && identifiedPinnedDocs.isEmpty() == false) {
+                pinnedQueryBuilder = new PinnedQueryBuilder(organicQuery, identifiedPinnedDocs.toArray(new Item[0]));
+                return pinnedQueryBuilder;
```

Comment:
Also, I think we want to ensure the pinned query builder has our `_name` & `boost`.

As for `organicQuery`, maybe it should as well.

Reply:
OK, looking at the behavior, I think organic should NOT inherit the name & boost. But, since the user does not directly supply the `PinnedQueryBuilder`, it may be assumed it should inherit the `_name` & `boost`. 

So, when we return the new `PinnedQueryBuilder` from this, it should have the name and boost applied. But, luckily, looking at `AbstractQueryBuilder#rewrite` its all handled for us. So, this and my previous comment are useless :)

Reply:
I'll add it for `pinnedQueryBuilder`. I had originally had it for `organicQuery` but @jimczi made a good point that the organic query could potentially have its own name and boost.